jax/experimental/sparse/linalg.py

```diff
@@ -0,0 +1,388 @@
+  else:
+    state = jax.lax.while_loop(cond, body, state)
+  i, X, _P, _R, _converged, theta = state
+
+  if debug:
+    return theta[0, :], X, i, diagnostics
+  return theta[0, :], X, i
+
+
+def _check_inputs(A, X):
+  n, k = X.shape
+  dt = X.dtype
+
+  if k == 0:
+    raise ValueError(f'must have search dim > 0, got {k}')
+
+  if k * 5 >= n:
+    raise ValueError(f'expected search dim * 5 < matrix dim (got {k * 5}, {n})')
+
+  test_output = A(jnp.zeros((n, 1), dtype=X.dtype))
+
+  if test_output.dtype != dt:
+    raise ValueError(
+        f'A, X must have same dtypes (were {test_output.dtype}, {dt})')
+
+  if test_output.shape != (n, 1):
+    s = test_output.shape
+    raise ValueError(f'A must be ({n}, {n}) matrix A, got output {s}')
+
+
+def _mm(a, b, precision=jax.lax.Precision.HIGHEST):
+  return jax.lax.dot(a, b, (precision, precision))
+
+def _generate_diagnostics(prev_XPR, X, P, R, theta, converged, adj_resid):
+  k = X.shape[1]
+  assert X.shape == P.shape
+
+  diagdiag = lambda x: jnp.diag(jnp.diag(x))
+  abserr = lambda x: jnp.abs(x).sum() / (k ** 2)
+
+  XTX = _mm(X.T, X)
+  DX = diagdiag(XTX)
+  orthX = abserr(XTX - DX)
+
+  PTP = _mm(P.T, P)
+  DP = diagdiag(PTP)
+  orthP = abserr(PTP - DP)
+
+  PX = abserr(X.T @ P)
+
+  prev_basis = prev_XPR.shape[1] - jnp.sum(jnp.all(prev_XPR == 0.0, axis=0))
+
+  return {
+      'basis rank': prev_basis,
+      'X zeros': jnp.sum(jnp.all(X == 0.0, axis=0)),
+      'P zeros': jnp.sum(jnp.all(P == 0.0, axis=0)),
+      'lambda history': theta[:k],
+      'residual history': jnp.linalg.norm(R, axis=0, ord=2),
+      'converged': converged,
+      'adjusted residual max': jnp.max(adj_resid),
+      'adjusted residual p50': jnp.median(adj_resid),
+      'adjusted residual min': jnp.min(adj_resid),
+      'X orth': orthX,
+      'P orth': orthP,
+      'P.X': PX}
+
+def _eigh_ascending(A):
+  w, V = jnp.linalg.eigh(A)
+  return w[::-1], V[:, ::-1]
+
+
+def _svqb(X):
+  # https://sdm.lbl.gov/~kewu/ps/45577.html
+  # May choose to truncate the basis rather than return
+  # a degenerate, non-orthogonal one.
+
+  norms = jnp.linalg.norm(X, ord=2, axis=0, keepdims=True)
+  X /= jnp.where(norms == 0, 1.0, norms)
+
+  inner = _mm(X.T, X)
+
+  w, V = _eigh_ascending(inner)
+
+  tau = jnp.finfo(X.dtype).eps * w[0]
+  padded = jnp.maximum(w, tau)
+  sqrted = jnp.where(tau > 0, padded, 1.0) ** (-0.5)
+  scaledV = V * sqrted[jnp.newaxis, :]
+  orthoX = _mm(X, scaledV)
+
+  keep = ((w > tau) * (jnp.diag(inner) > 0.0))[jnp.newaxis, :]
+  orthoX *= keep
+  norms = jnp.linalg.norm(orthoX, ord=2, axis=0, keepdims=True)
+  keep *= norms > 0.0
+  orthoX /= jnp.where(keep, norms, 1.0)
+  return orthoX
+
+
+def _project_out(basis, U):
+  # "twice is enough" from shoyer's reference:
+  # http://slepc.upv.es/documentation/reports/str1.pdf
+
+  for _ in range(2):
+    U -= _mm(basis, _mm(basis.T, U))
+    U = _orthonormalize(U)
+
+  # It's crucial to end on a subtraction of the original basis.
```

Comment:
Hmm this is very interesting. If you allow these small non-zero components in the complement of the span of the large column to be normalized, it can act as a kind on randomized restarting, but I agree that it is difficult to control. I used it in Lanczos (PROPACK) to recover from computing an invariant subspace before convergence, but it is tricky.

Reply:
Indeed! I hope to find time to at least dig into it to do an internal presentation on it, it seemed like a nice new learning from implementing this.

One approach that also worked, but had slightly worse quality, was to leave these "random restarts" in and do another svqb() to orthonormalize the Ritz basis (this is also more expensive b/c the whole 3k x 3k system needs to be mutually orthonormalized).